rfcs/2020-11-06-4909-remap-template-fields.md

```diff
@@ -0,0 +1,211 @@
+# RFC 4909 - 2020-11-06 - Remap support for template strings
+
+We would like to unify the templating syntax within configuration fields with
+the Remap language.
+
+## Scope
+
+This RFC will look into ways we can use Remap whilst still supporting the
+current method of templating fields (known throughout this RFC as the template
+syntax).
+
+## Motivation
+
+Using the Remap language rather than the template syntax provides much greater
+power in how these fields are defined.
+
+The advantages of using Remap rather than the existing template syntax are:
+
+- One familiar syntax and function reference for Vector.
+- Access to all of remap's functions for templating.
+
+However, we do still need to support the template syntax in order to maintain
+backward compatability.
+
+## Internal Proposal
+
+There are two issues that need resolving to allow Remap to be used.
+
+### Use distinct syntax to distinguish between template and Remap syntax
+
+We will specify that double parentheses (`{{..}}`) are used for template
+syntax (the current technique) and triple parentheses (`{{{...}}}`) are used
+for Remap syntax.
+
+This would provide a clear and unambiguous way to distinguish between the
+syntaxes rather than rely on a set of heuristics.
+
+The script will need to return a value that can resolve to a String. `to_string`
+will be called on the final result. `String`, `Integer`, `Bool`, `Timestamp`
+will be allowed results. The script will, as much as possible, resolve this at
+loadtime. However, as event field types can resolve any type there will be some
+limitations to how much validation can be done.
+
+### Removing fields used in the template
+
+A feature available with the template syntax is to be able to list all the
+fields used within the template.
+
+The Loki sink wants a list of the fields that were used in the
+template. It uses this list to remove these fields from the event sent to Loki.
+
+If you had this sink:
+
+```toml
+[sinks.loki]
+  type = "loki"
+  inputs = ["json"]
+  endpoint = "http://localhost:3100"
+  remove_label_fields = true
+  labels.key = "{{foo}}-{{buzz}}"
+```
+
+And this event was sent to it:
+
+```json
+{"foo": "bar", "buzz": "zab", "message1": "thing", "message2": "thong"}
+```
+
+The actual message sent to Loki would be:
+
+```json
+{"message1": "thing", "message2": "thong"}
+```
+
+With the label `key = bar-zab` attached to the message. The fields `foo` and
+`buzz` that were used in defining the label have been removed from the message.
+
+It is not going to be possible to do this with the Remap syntax as it is much
+more complicated and ambiguous to determine which fields are used to generate
+the templated string.
+
+### Add `remap.before` and `remap.after` fields
+
+To replicate this functionality we will add `remap.before` and `remap.after`
```

Comment:
So, I've thought about `remap.before` and `remap.after` a bit more, and looking at this example, it's not so clear what `remap.after` means in _this_ context, and how it would relate to the same field in _other_ components.

For the `loki` sink, to have `remap.after` be useful, it would have to fire _before_ the data is sent to Loki, but _after_ the `labels.key` is applied.

However, for transforms or sources, `remap.after` would just run "right before the transform/source is done with the event".

So, I guess `remap.after` is described as

> Runs _before_ the end of the component. For **sources** and **transforms** this is _before_ the event is passed to the next component, for **sinks** this is _before_ the event is delivered by Vector to the configured sink.

For `remap.before`, it would be similar, except that sources/sinks are swapped:

> Runs right _after_ the start of the component. For **transforms** and **sinks** this is _before_ the event is processed by the given component, for **sources** this is _after_ the event is received by Vector, but before the source has processed the event.

It requires a bit of explaining, but it does provide the most amount of flexibility.

Reply:
:+1: I think this is the most easily understandable way to model it.

Reply:
It's not entirely clear to me why we need two variants. For sources, it really only makes sense to modify the event after we've received it, and likewise, for sinks, we can only change it before we send it.

Transforms do have a before and after, but it seems confusing to wrap one transform with what's effectively just another transform. Do we have a good example of a transform that'd benefit from being able to have pre- or post-processing with remap without just chaining a remap transform? It seems to me that a single `remap` option on sources and sinks should cover the vast majority of use cases.

Coming back to templating specifically, I'm not sure it makes sense to encourage users to use multiple different remap executions (one generic pre-sink and one per templated field) and try to explain their ordering. Why not just have one remap execution that happens first and leave the templating as simple field access?

Initially, we can just leave the options to remove templated fields as-is, but eventually, with the introduction of better event metadata, we could give remap access to that metadata and allow users to remove fields from the normal event data while still having access via metadata for templated fields.

I understand the desire to use remap for literally everything, but there is a reason that most templating solutions use a dedicated language.

Reply:
Agreed, there are considerations to be made here. We agreed that this RFC is not the place to introduce `remap.before/after` or any other variant.

I'm sure whoever picks up that RFC work can expand on the use-case for before/after (or lack thereof).